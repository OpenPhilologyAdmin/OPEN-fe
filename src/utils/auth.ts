import { COOKIES } from "@/constants/cookies";
import clientCookie from "js-cookie";

export const getAccessToken = (): string | undefined => {
  return clientCookie.get(COOKIES.ACCESS_TOKEN);
};

export const setAccessToken = (token: string) => {
  clientCookie.set(COOKIES.ACCESS_TOKEN, token);
};

export const removeAccessToken = () => {
  clientCookie.remove(COOKIES.ACCESS_TOKEN);
};
