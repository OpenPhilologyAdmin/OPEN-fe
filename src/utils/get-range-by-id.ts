export const getRangeById = <ElementType extends { id: number }>(
  elements: ElementType[],
  startElementId: number,
  endElementId: number,
) => {
  const elementIds = elements.map(element => element.id);
  const startElementIndex = elementIds.indexOf(startElementId);
  const endElementIndex = elementIds.indexOf(endElementId);

  // only one element followed by apparatus index was selected
  if (endElementId === Infinity) {
    return [elements[startElementIndex]];
  }

  if (startElementId === Infinity) {
    return [elements[endElementIndex]];
  }

  return startElementIndex < endElementIndex
    ? elements.slice(startElementIndex, endElementIndex + 1)
    : elements.slice(endElementIndex, startElementIndex + 1);
};
