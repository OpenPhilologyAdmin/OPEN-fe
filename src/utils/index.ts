// placeholder file for structure
export const placeholder = "";
