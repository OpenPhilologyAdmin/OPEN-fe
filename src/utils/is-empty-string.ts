export const isEmptyString = (str?: string) => str === "";
