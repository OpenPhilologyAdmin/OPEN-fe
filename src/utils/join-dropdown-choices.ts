import { NormalizedDropdownOption } from "@/components/dropdown/item";

export function joinDropdownChoices(options: NormalizedDropdownOption[], prompt: string): string {
  const selectedOptions = options.filter(option => Boolean(option.selected));

  if (!(selectedOptions.length > 0)) return prompt;

  return selectedOptions.map(option => option.label).join(", ");
}
