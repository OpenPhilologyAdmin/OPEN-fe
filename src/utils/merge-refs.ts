export function mergeRefs<GenericElement extends HTMLElement | null = HTMLDivElement>(
  ...refs: (React.MutableRefObject<GenericElement> | React.Ref<GenericElement> | undefined)[]
) {
  return (value: GenericElement) => {
    refs.forEach(ref => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref) {
        (ref as React.MutableRefObject<GenericElement>).current = value;
      }
    });
  };
}
