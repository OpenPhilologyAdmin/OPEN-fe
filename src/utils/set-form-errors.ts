import { ErrorOption } from "react-hook-form";

type ValueOf<AnyType> = AnyType[keyof AnyType];

type SetFormErrorsParams<FieldValues> = {
  apiError: API.Error;
  fields: FieldValues;
  setError: (field: ValueOf<FieldValues>, errorOptions: ErrorOption) => void;
};

export const setFormErrors = <FieldValues>({
  apiError,
  fields,
  setError,
}: SetFormErrorsParams<FieldValues>) => {
  Object.values(fields).forEach(field => {
    if (apiError[field]) {
      setError(field, { message: apiError[field][0] });
    }
  });
};
