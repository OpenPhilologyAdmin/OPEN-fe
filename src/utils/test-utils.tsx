import { ReactNode } from "react";
import ModalMock from "react-modal";
import { QueryClientProvider } from "react-query";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { NextRouter } from "next/router";

import { ToastProvider } from "@/components/ui/toast";
import { UserProvider } from "@/contexts/user";
import { mswServer } from "@/mocks/index";
import { mockQueryCache, mockQueryClient } from "@/mocks/mock-query";
import { mockRouter } from "@/mocks/mock-router";
import { mockUser } from "@/mocks/mock-user";
import * as testing from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "src/contexts/theme";

beforeAll(() => {
  mockQueryCache.clear();
  mswServer.listen({
    onUnhandledRequest: "error",
  });
});

afterEach(() => {
  mswServer.resetHandlers();
  mockQueryCache.clear();
});

afterAll(() => mswServer.close());

function MockProvider({ children, user }: { children: ReactNode; user?: API.User }) {
  return (
    <ThemeProvider initialTheme="LIGHT">
      <UserProvider initialUser={user}>
        <QueryClientProvider client={mockQueryClient}>{children}</QueryClientProvider>
      </UserProvider>
    </ThemeProvider>
  );
}

/**
 * Include this provider in tests that rely on react-tostify toast component
 *
 * This is not mocked globally because the way the library is written, it would disrupt all existing snapshots by injecting the provider
 */
export function MockToastProvider() {
  return <ToastProvider />;
}

type DefaultParams = Parameters<typeof testing.render>;
type RenderUI = DefaultParams[0];
type RenderOptions = Omit<DefaultParams[1], "queries"> & {
  router?: Partial<NextRouter>;
  user?: Partial<API.User>;
};

/**
 * Custom render function that wraps the default render from @testing-library/react
 *
 * This function is used to wrap components rendered in tests with mandatory providers used in the app
 *
 * @param ui The rendered component
 * @param options Default render options extended by mocked NextRouter
 * @example render(<Component />, { router: { query: { token: 123 } } })
 */
const customRender = (ui: RenderUI, options?: RenderOptions) => {
  const wrapper = ({ children }: { children: ReactNode }) => (
    <MockProvider user={{ ...mockUser, ...options?.user }}>
      {options?.router ? (
        <RouterContext.Provider value={{ ...mockRouter, ...options.router }}>
          {children}
        </RouterContext.Provider>
      ) : (
        children
      )}
    </MockProvider>
  );

  return testing.render(ui, { wrapper, ...options });
};

// Mock external modal library
const oldFn = ModalMock.setAppElement;

ModalMock.setAppElement = element => {
  if (element === "#__next") {
    // otherwise it will throw aria warnings.
    return oldFn(document.createElement("div"));
  }

  oldFn(element);
};

// ! We want to export all from testing library and override custom render on purpose
// eslint-disable-next-line import/export
export * from "@testing-library/react";
// eslint-disable-next-line import/export
export { customRender as render, mswServer as mockServer, userEvent };
