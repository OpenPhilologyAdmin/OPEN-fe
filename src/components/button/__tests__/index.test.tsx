import { render, screen } from "@/utils/test-utils";

import Button from "..";

const text = "Some text";

describe("Button", () => {
  it("renders a primary button unchanged", () => {
    const { container } = render(<Button variant="primary" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a secondary button unchanged", () => {
    const { container } = render(<Button variant="secondary" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a primary-outline button unchanged", () => {
    const { container } = render(<Button variant="primary-outline" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a secondary-outline button unchanged", () => {
    const { container } = render(<Button variant="secondary-outline" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a primary-ghost button unchanged", () => {
    const { container } = render(<Button variant="primary-ghost" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a secondary-ghost button unchanged", () => {
    const { container } = render(<Button variant="secondary-ghost" />);

    expect(container).toMatchSnapshot();
  });

  it("renders a button with left prop correctly", () => {
    render(<Button variant="primary" left={<div>{text}</div>} />);

    expect(screen.getByText(text)).toBeInTheDocument();
  });

  it("renders a button with right prop correctly", () => {
    render(<Button variant="primary" right={<div>{text}</div>} />);

    expect(screen.getByText(text)).toBeInTheDocument();
  });

  it("renders a button with disabled prop correctly", () => {
    render(<Button variant="primary" disabled />);

    expect(screen.getByRole("button")).toBeDisabled();
  });

  it("renders a button with small prop correctly", () => {
    const { container } = render(<Button small />);

    expect(container).toMatchSnapshot();
  });

  it("renders a button as link", () => {
    const href = "https://example.com";

    render(
      <Button variant="primary" href={href}>
        {text}
      </Button>,
    );

    expect(screen.getByRole("link", { name: text }).closest("a")).toHaveAttribute("href", href);
  });
});
