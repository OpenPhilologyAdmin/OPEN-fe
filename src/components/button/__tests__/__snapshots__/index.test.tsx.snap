// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a button with small prop correctly 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 32px;
  border-radius: 4px;
  padding: 6px 12px;
  border-color: #5D2DEF;
  background-color: #5D2DEF;
  color: #FFFFFF;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a primary button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 24px;
  border-color: #5D2DEF;
  background-color: #5D2DEF;
  color: #FFFFFF;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a primary-ghost button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 0px;
  border-color: transparent;
  background-color: transparent;
  color: #5D2DEF;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a primary-outline button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 24px;
  border-color: #5D2DEF;
  background-color: #FFFFFF;
  color: #5D2DEF;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a secondary button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 24px;
  border-color: #FC6145;
  background-color: #FC6145;
  color: #FFFFFF;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a secondary-ghost button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 0px;
  border-color: transparent;
  background-color: transparent;
  color: #FC6145;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;

exports[`Button renders a secondary-outline button unchanged 1`] = `
.c1 {
  font-family: "Inter";
  font-style: normal;
  font-weight: 600;
  font-size: 16px;
  line-height: 24px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0;
  border: 2px solid;
  outline: none;
  max-height: 48px;
  border-radius: 4px;
  padding: 12px 24px;
  border-color: #FC6145;
  background-color: #FFFFFF;
  color: #FC6145;
  cursor: pointer;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;
